### PR TITLE
SDK: Refactor StateChangeOp and add path parameter

### PR DESCRIPTION
**Pull request by rbx**
Refactors StateChangeOp to allow concurrent invocations and adds a path parameter to the call that allows to target subsets of the topology.

### CODE REVIEW

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/sdk/Topology.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5176147&bid=16989917)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/sdk/Topology.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5176147&bid=16989917)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/sdk/Topology.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5176147&bid=16989917)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/sdk/Topology.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5176147&bid=16989917)

**Comment by rbx** [comment]
The remaining test failures seem unrelated.

File: fairmq/sdk/Topology.h [hunk 1]
```diff
@@ -352,85 +506,118 @@ class BasicTopology : public AsioBase<Executor, Allocator>
...
+                std::forward_as_tuple(id),
+                std::forward_as_tuple(id,
+                                        transition,
+                                        fDDSTopo.GetTasks(path),
+                                        fStateData,
+                                        timeout,
+                                        fMtx,
+                                        AsioBase<Executor, Allocator>::GetExecutor(),
+                                        AsioBase<Executor, Allocator>::GetAllocator(),
+                                        std::move(handler)));
```
Comment (dennisklein) [inline fairmq/sdk/Topology.h]:
```suggestion
                                      transition,
                                      fDDSTopo.GetTasks(path),
                                      fStateData,
                                      timeout,
                                      fMtx,
                                      AsioBase<Executor, Allocator>::GetExecutor(),
                                      AsioBase<Executor, Allocator>::GetAllocator(),
                                      std::move(handler)));
```

File: fairmq/sdk/Topology.h [hunk 2]
```diff
@@ -523,6 +706,7 @@ class BasicTopology : public AsioBase<Executor, Allocator>
         {
             if (!fOp.IsCompleted() && fCount == fTasks.size()) {
                 fTimer.cancel();
+                // TODO: this should be initiated differently
```
Comment (dennisklein) [inline fairmq/sdk/Topology.h]: ?

File: fairmq/sdk/Topology.h [hunk 3]
```diff
@@ -270,15 +270,173 @@ class BasicTopology : public AsioBase<Executor, Allocator>
...
+            if (!fOp.IsCompleted() && fCount == fTasks.size()) {
+                Complete(std::error_code());
+            }
+        }
+
+        /// precondition: fMtx is locked.
+        auto Complete(std::error_code ec) -> void
+        {
+            fTimer.cancel();
+            // TODO: this should be initiated differently
```
Comment (dennisklein) [inline fairmq/sdk/Topology.h]: ?

File: fairmq/sdk/Topology.h (hunk 2 above)
Comment (rbx) [reply fairmq/sdk/Topology.h]: The `TryCompletion` method is called from `asio::async_initiate` directly and, if it can complete, it will complete the Op without directly. In the previous version we had special handling for thise case, at least for the StateChange Op. We discussed that we should also do it for the new Ops, but this is not yet implemented in this PR. I'll rephrase it to be more clear and move the comment to the place where TryCompletion is called.

Comment (rbx) [reply fairmq/sdk/Topology.h]: Updated/moved the comments inside the `async_initiate`.

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/sdk/Topology.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5176147&bid=16989917)

**Comment by rbx** [comment]
One of the Macs is down it seems.

**Comment by dennisklein** [comment]
> One of the Macs is down it seems.

Just busy with some FairSoft CI jobs atm.
